src/widget/wfindonwebmenu.cpp

```diff
@@ -0,0 +1,167 @@
+            return kSearchUrlLastFmArtist;
+        } else {
+            return kSearchUrlLastFmAlbum;
+        }
+    case WFindOnWebMenu::Service::SoundCloud:
+        if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Title ||
+                trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::ArtistAndTitle) {
+            return kSearchUrlSoundCloudTitle;
+        } else if (trackSearchProperty == WFindOnWebMenu::TrackSearchProperties::Artist) {
+            return kSearchUrlSoundCloudArtist;
+        } else {
+            return kSearchUrlSoundCloudAlbum;
+        }
+    default:
+        return kSearchUrlDefault;
+    }
+}
+} // namespace
+
+WFindOnWebMenu::WFindOnWebMenu(QWidget* parent)
+        : QMenu(tr("Find on Web"), parent) {
+}
+
+void WFindOnWebMenu::addSubmenusForServices(const Track& track) {
+    WFindOnWebMenu::populateFromTrackProperties(
+            track,
+            QString(tr("SoundCloud")),
+            Service::SoundCloud);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("LastFm")), Service::LastFm);
+    WFindOnWebMenu::populateFromTrackProperties(
+            track, QString(tr("Discogs")), Service::Discogs);
+}
+
+void WFindOnWebMenu::addActions(Service service,
+        const QString& trackProperty,
+        QMenu* pServiceMenu,
+        TrackSearchProperties trackSearchProperty) {
+    const QString prefixAction = composePrefixAction(trackSearchProperty);
+    pServiceMenu->addAction(
+            composeActionText(prefixAction, trackProperty),
+            this,
+            [this, service, trackSearchProperty, trackProperty]() {
+                emit triggerBrowser(service, trackSearchProperty, trackProperty);
+            });
+}
+
+void WFindOnWebMenu::openInBrowser(Service service,
+        TrackSearchProperties trackSearchProperty,
+        const QString& query) {
+    const QString serviceUrl = getServiceUrl(service, trackSearchProperty);
+    QUrlQuery urlQuery;
+    urlQuery.addQueryItem("q", query);
+    QUrl url(serviceUrl);
+    url.setQuery(urlQuery);
+    VERIFY_OR_DEBUG_ASSERT(!url.isEmpty()) {
+        qWarning() << "URL has no data!";
+        return;
+    }
+    QDesktopServices::openUrl(url);
+}
+
+void WFindOnWebMenu::populateFromTrackProperties(
+        const Track& track,
+        const QString& serviceTitle,
+        Service service) {
+    const auto artist = track.getArtist();
+    auto pServiceMenu = make_parented<QMenu>(this);
+    pServiceMenu->setTitle(serviceTitle);
+    addMenu(pServiceMenu);
+    addSeparator();
+    if (!artist.isEmpty()) {
+        addActions(service,
+                artist,
+                pServiceMenu,
+                WFindOnWebMenu::TrackSearchProperties::Artist);
+    }
+    {
+        const auto trackTitle = track.getTitle();
+        const auto artistWithTrackTitle = composeSearchQuery(trackTitle, artist);
```

Comment:
It is confusing to compute a value that might get discarded by the following if statement. Abort early.

Reply:
Solved :+1: